### ISSUE TITLE
Add more patterns/examples for deploys

### ISSUE THREAD

**Issue by mkinney** [Fizzadar/pyinfra#200]
I would like to see more patterns/examples for deploys.

1) Ansible's docs usually contain a useable snippet as a starting point. Could the pyinfra docs be expanded to include examples? For example, the apt.packages could have an example of what a typical code snippet would look like.

2) Would like to see examples on using the "when" or a conditional task, so as to make deployments quicker/idempotent.

For example, the "if CentOS" in deploys.rst could show how to also use the "when" format.

Another example would be to check for the existence of a file then use that as a "when" condition.

Yet another example, would be to run a remote shell command and use the output from that as a "when" condition in a task. (a good example of this would be check if docker apt key is installed before needing to download that key)

**Comment by Fizzadar** [comment]
For 1 I think this is a fantastic idea - I think adding examples to each operartions docstring would be a huge improvement to the documentation especially people new to pyinfra - I think the current list of arguments can sometimes be quite overwhelming, especially when most are often not required.

As for 2 - I think the `when` argument may be worth deprecating/removing entirely - and normal Python conditionals should be used; ie:

```py
if host.fact.linux_name == 'CentOS':
   dnf.packages(...)
```

IMO this is easier to read (any more Python-y) than the old when syntax (which dates back to pyinfra not handling conditionals properly):

```py
dnf.packages(
    ...
    when=host.fact.linux_name=='CentOS',
)
```

For a file check this can already be achieved with:

```py
if host.fact.file('/path/to/file') is None:
    files.put(...)
```

And for the apt key example - ideally the `apt.key` operation should handle this internally (it does not currently). Ie always calling `apt.key(...)` in the operation and pyinfra figures out whether the key exists, and if so does nothing. This is how most modules work where possible (ie `apt.packages('iftop')` will do nothing if `iftop` is already installed).

There is also a [documented example](https://pyinfra.readthedocs.io/en/latest/examples/dynamic_execution_deploy.html) for dynamic execution of commands during a deploy.

**Comment by mkinney** [comment]
I agree on the "when" part. Thanks for sharing the "if" with the "file" fact. I'll leverage that for stuff in the future.

It is funny that I try hard to find an example where a "system" command is needed and you show that it should be handled internally. (which I agree, but I'm just providing an example that I found... lol)

So, I took a stab at the apk doc. See it here:
https://github.com/Fizzadar/pyinfra/compare/master...mkinney:add-doc-examples?expand=1

A few questions:
1) Why does that show all of my prior commits? I thought it would just go from master and show the two commits. I wonder if my refresh upstream commands of:

```
git fetch upstream
git checkout master
git merge upstream/master
git push
```

Is missing some step. Any ideas?

2) I also added and "examples" (perhaps it should be "standalone_examples" or something) and added that same code snippet. That is duplicated effort. Any way to say... go include that code snippet in that dir from inside the .rst file?

3) How do you view the .rst when doing dev? I am using MacView, but it does not render correctly?

4) Might be answered in the answer for above, but how do you locally generate the doc?

**Comment by mkinney** [comment]
5. Do you want to include this pattern?

https://github.com/hoh/pyinfra-formulas/blob/master/formulas/nodejs/__init__.py#L33


6. I also need some help with processing a simple command. For instance, if I have something like this:

`server.shell('apt-key list')`

How do I get the stdout?

**Comment by mkinney** [comment]
7. Is there a way to get *all* facts?

8. What is the list of the meta info from a command (like `changed`)?

**Comment by Fizzadar** [comment]
@mkinney the APK docs look good - one thing is the `docs/modules/*.rst` are _autogenerated_ so you'll want to add the example in the Python module itself; ie under https://github.com/Fizzadar/pyinfra/blob/master/pyinfra/modules/apk.py#L48. (On that note I'm going to investigate _not_ having the autogenerated docs in git, and generated on doc build).

I've answered your questions below:

1. That should be fixed now - I think it was pre my merge of the Mech branch?

2. It is possible but with the above change it won't be.. it might be worth putting the APK example in `examples/packages.py`. Basically the idea is the examples directory is full deploys, whereas examples for single operations would live in the Python docstrings.

3 & 4. If you have everything installed in a virtualenv you should be able to run `scripts/build_docs.sh` to generate all the docs, and then load up `docs/build/index.html` in a browser.

5 & 8. Yep! The meta output can be really helpful here; and is very briefly [documented here](https://pyinfra.readthedocs.io/en/latest/deploys.html#operation-meta) - the implementation is @ https://github.com/Fizzadar/pyinfra/blob/master/pyinfra/api/operation.py#L44 - basically keys are `changed`, `commands` and `hash` (currently).

6. This is not currently possible - but if you use `host.fact.command('apt-key list')` that will return the stdout.

7. Not currently! They can be listed (`pyinfra --facts`) but no way to fetch them all as is.

Let me know if you've any more questions - this is _incredibly_ helpful for improving the pyinfra documentation and massively appreciated!

**Comment by mkinney** [comment]
Thanks for looking at 1 above.

I tried (but failed) to get the doc updated. The scripts work, but I could not get the locally built files to reflect changes. I was trying to use:

```
    Example:
    .. literalinclude:: ../../examples/apt.py
```

I think I'm just going to build some standalone examples for now. Let me know if you figure out how to get code snippets to be included.

FYI, I did something like this to get all facts:

`pyinfra 192.168.2.228 --user vagrant --password vagrant --facts | tail -n +2 | awk '{ tnf=split($0,x); for (i=0;i < tnf;i++) { s=x[i]; sub(",", "", s); all[s]=1 }}; END { for (e in all) { if (e != "") { print e }}}' | sort | awk '{ printf("%s ", $1) } END { print ""}'`

Which returned:
`apk_packages apt_sources block_devices brew_casks brew_packages command cpus date deb_package deb_packages directory docker_container docker_containers docker_image docker_images docker_network docker_networks docker_single_mixin file find_directories find_files find_in_file gem_packages git_branch git_config home initd_status ip6tables_chains ip6tables_rules iptables_chains iptables_rules ipv4_addresses launchd_status link linux_distribution linux_name lsb_release memory mounts mysql_databases mysql_user_grants network_devices openvz_containers os pacman_packages pip_packages pkg_packages postgresql_databases rcd_status rpm_package sha1_file short_fact_base socket sysctl systemd_enabled upstart_status`

Then I removed the ones that required args:
`pyinfra 192.168.2.228 --user vagrant --password vagrant fact apk_packages apt_sources block_devices brew_casks brew_packages cpus date deb_packages docker_containers docker_images docker_networks gem_packages git_config home initd_status ip6tables_chains ip6tables_rules iptables_chains iptables_rules ipv4_addresses launchd_status linux_distribution linux_name lsb_release memory mounts network_devices openvz_containers os pacman_packages pip_packages pkg_packages postgresql_databases rcd_status sysctl systemd_enabled upstart_status`

The reason for this is so I could look for any fact that might contain a string (like "amd64").

8. init was not obvious to me that it was for services. Not sure how to show that in the docs.

**Comment by Fizzadar** [comment]
Hmm - for the docs build what error were you getting? I'll try a fresh virtualenv later to see if I can replicate.

For the facts, I believe the [`arch` fact](https://pyinfra.readthedocs.io/en/latest/facts/server.html#arch) would give you the amd64 string. I've also added [an issue to get all facts in one go](https://github.com/Fizzadar/pyinfra/issues/203) as it'd be nice to have.

8 - interesting, I think this makes sense. `service.systemd` vs `init.systemd` I suppose. v1 is coming up which could break this; or `service` could be an alternative name for the same module. Will have a think about this!

**Comment by Fizzadar** [comment]
I've added an example for documentation-based examples here: https://github.com/Fizzadar/pyinfra/commit/12bf145ec0971a578546deeccae83cca79ab1da1 which can be seen here: https://pyinfra.readthedocs.io/en/latest/modules/apk.html#apk-packages.

Keen to keep the inline examples as short as possible (no imports/etc) or the docs will get out of hand. This also nicely separates `examples/` vs the docs.

**Comment by mkinney** [comment]
Nice! That looks great. Thanks.

I still think it would be good to have a "standalone scripts" for each in examples somewhere. That way, people don't get tripped up on what the includes would need to be or anything. Plus they could show various patterns.

**Comment by Fizzadar** [comment]
Agreed - it'd be neat to have examples that cover using every operation in a given module. It's also much more explicit than some of the current example names!

**Comment by mkinney** [comment]
Working on it... it is slow going, but I think that's ok. I've found a few bugs and will probably need your help with some things. (as evidenced by some new issues)

See https://github.com/Fizzadar/pyinfra/compare/master...mkinney:add-examples?expand=1 for progress, if you're curious.

**Comment by Fizzadar** [comment]
Just had a quick read through - looking great so far 👍

**Comment by mkinney** [comment]
I was unable to test pacman as the Arch instance I was using kept erroring on gpg keys. I've asked the maintainer wouldn't mind re-spinning the image. (https://github.com/mrlesmithjr/packer-templates/issues/54)

I was unable to test pkg.packages() as neither freeBSD 11 nor 12 would work. (errors were something like "pkg_info" and probably should be "pkg info"). I tried openbsd, but it would take longer than 10 seconds to connect. I tried hacking the connection timeout, but ran into a paramiko error. So, not sure if you have any better luck with validating pkg.packages().
